scio-bigquery/src/main/scala/com/spotify/scio/bigquery/nio/BigQueryIO.scala

```diff
@@ -0,0 +1,414 @@
+  * @group input
+  */
+final case class Select(sqlQuery: String) extends ScioIO[TableRow] {
+  import Select._
+  type ReadP = ReadParam
+  type WriteP = Nothing // ReadOnly
+
+  def id: String = sqlQuery
+
+  def read(sc: ScioContext, params: ReadParam): SCollection[TableRow] =
+    params match {
+      case FlattenResults(flattenResults) =>
+        Reads.bqReadQuery(sc)(bqio.BigQueryIO.readTableRows(), sqlQuery, flattenResults)
+    }
+
+  private lazy val bqc = BigQueryClient.defaultInstance()
+
+  def tap(params: ReadParam): Tap[TableRow] =
+    params match {
+        case FlattenResults(f) =>
+          val table = bqc.query(sqlQuery, flattenResults = f)
+          BigQueryTap(table)
+    }
+
+  def write(data: SCollection[TableRow], params: WriteP): Future[Tap[TableRow]] =
+    throw new IllegalStateException("Select queries are read-only")
+}
+
+object Select {
+  sealed trait ReadParam
+  final case class FlattenResults(value: Boolean = false) extends ReadParam
+}
+
+/**
+ * Get an IO for a BigQuery table.
+ * @group input
+ */
+final case class TableRef(table: TableReference) extends ScioIO[TableRow] {
+  type ReadP = Unit
+  type WriteP = TableRef.WriteParam
+
+  def id: String = bqio.BigQueryHelpers.toTableSpec(table)
+
+  def read(sc: ScioContext, params: ReadP): SCollection[TableRow] =
+    Reads.bqReadTable(sc)(bqio.BigQueryIO.readTableRows(), table)
+
+  def tap(read: ReadP): Tap[TableRow] =
+    BigQueryTap(table)
+
+  def write(data: SCollection[TableRow], params: WriteP): Future[Tap[TableRow]] =
+    params match {
+      case TableRef.Parameters(schema, writeDisposition, createDisposition, tableDescription) =>
+        var transform = bqio.BigQueryIO.writeTableRows().to(table)
+        if (schema != null) {
+          transform = transform.withSchema(schema)
+        }
+        if (createDisposition != null) {
+          transform = transform.withCreateDisposition(createDisposition)
+        }
+        if (writeDisposition != null) {
+          transform = transform.withWriteDisposition(writeDisposition)
+        }
+        if (tableDescription != null) {
+          transform = transform.withTableDescription(tableDescription)
+        }
+        data.applyInternal(transform)
+
+        if (writeDisposition == WriteDisposition.WRITE_APPEND) {
+          Future.failed(new NotImplementedError("BigQuery future with append not implemented"))
+        } else {
+          data.context.makeFuture(BigQueryTap(table))
+        }
+    }
+}
+
+object TableRef {
+  sealed trait WriteParam
+  final case class Parameters(
+    schema: TableSchema,
+    writeDisposition: WriteDisposition,
+    createDisposition: CreateDisposition,
+    tableDescription: String) extends WriteParam
+}
+
+/**
+ * Get an SCollection for a BigQuery table.
+ * @group input
+ */
+final case class TableSpec(tableSpec: String) extends ScioIO[TableRow] {
+  type ReadP = Unit
+  type WriteP = TableSpec.WriteParam
```

Comment:
use `TableRef.WriteParam` type? don't see why we will have different write param support for `TableRef` and `TableSpec`. I could be wrong though.

Reply:
To expose a coherent API to the end user. The parameters for xxxIO should always be found in the xxxIO object.

Reply:
Yeah, that makes sense 👍